### PR TITLE
Adding complex dFe cycling to WOMBAT-lite

### PR DESCRIPTION
**Pull request by pearseb**
I've added a complex dFe chemistry and cycling component to WOMBAT-lite.

This includes:
-  the addition of 3 tracers: phyfe, zoofe, and detfe.
- explicit consideration of dFe scavenging on biogenic particles, nanoparticle precipitation when dFe concentrations are very high, and coagulation of colloidal dFe onto organic detritus.
- variable Fe:C ratios of phytoplankton via a quota model, with luxury dFe uptake
- variable Fe:C ratios of zooplankton and detritus, importantly increasing Fe:C ratios of detritus with depth as per observations.
- pH and temperature and salinity dependence of free dFe
- sedimentary dFe pool that accumulates and remineralises independently of the carbon pool.

Current dFe output looks okay, but there is too little at the surface and dFe limitation of phytoplankton is too extensive at this stage. The reason for this is because I need to make additional updates to the zooplankton cycling of nutrients, which will come in a subsequent step of development on a new branch.

Current run time has increased from ~27 min to ~34 minutes per model year (25%) with the addition of the 3 tracers and additional computations.

======================================================================================
                  Resource Usage on 2024-10-29 16:51:55:
   Job Id:             127813476.gadi-pbs
   Project:            p66
   Exit Status:        0
   Service Units:      332.96
   NCPUs Requested:    288                    NCPUs Used: 288    
                                           CPU Time Used: 137:32:56    
   Memory Requested:   1000.0GB              Memory Used: 167.34GB    
   Walltime requested: 10:00:00            Walltime Used: 00:34:41    
   JobFS requested:    600.0MB                JobFS used: 8.16MB    
======================================================================================

### CODE REVIEW

File: generic_tracers/generic_WOMBATlite.F90 [hunk 1]
```diff
@@ -1477,6 +1705,13 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
     real, dimension(:), allocatable         :: par_mid
     real, dimension(:,:,:), allocatable     :: par_phy, par_phymld
     real, dimension(4,61)                   :: zbgr
+    real                                    :: ztemk, fe_keq, fe_par, fe_sfe, fe_tfe 
+    real                                    :: fe_III, fe_lig, partic, fe_col
+    real                                    :: fesol1, fesol2, fesol3, fesol4, fesol5, hp, fe3sol, precip
```
Comment (dougiesquire) [inline generic_tracers/generic_WOMBATlite.F90:1710]: `precip` doesn't seem to be used anywhere

File: generic_tracers/generic_WOMBATlite.F90 [hunk 2]
```diff
@@ -1989,126 +2277,275 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+
+      ! Determine equilibriuim fractionation of the remain dFe (non-colloidal fraction) into Fe' and L-Fe
+      fe_keq = 10**( 17.27 - 1565.7 / ztemk ) * 1e-9 ! Temperature reduces solubility
+      fe_par = 4.77e-7 * wombat%radbio3d(i,j,k) * 0.5 / 10**(-6.3) ! Light increases solubility
+      fe_sfe = max(0.0, biofer - fe_col)
+      fe_tfe = (1.0 + fe_par) * fe_sfe
+      fe_III = ( -( 1. + wombat%ligand * fe_keq + fe_par - fe_sfe * fe_keq ) &
+                 + SQRT( ( 1. + wombat%ligand * fe_keq + fe_par - fe_sfe * fe_keq )**2 &
+                         + 4. * fe_tfe * fe_keq) ) / ( 2. * fe_keq + epsi )
+      fe_lig = max(0.0, fe_sfe - fe_III)
```
Comment (dougiesquire) [inline generic_tracers/generic_WOMBATlite.F90:2333]: `fe_lig` doesn't seem to be used anywhere?

File: generic_tracers/generic_WOMBATlite.F90 [hunk 3]
```diff
@@ -1461,9 +1689,9 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
     real, dimension(nbands)                 :: sw_pen
     real                                    :: swpar
     real                                    :: u_npz, g_npz
-    real                                    :: biophy, biono3, biofer
+    real                                    :: biophy, biozoo, biodet, biono3, biofer, biocaco3
+    real                                    :: biophyfe, biozoofe, biodetfe
```
Comment (dougiesquire) [inline generic_tracers/generic_WOMBATlite.F90:1693]: `biozoofe` and `biodetfe` don't seem to be used anywhere?

**Review by dougiesquire — COMMENTED** [review]
Nice @pearseb. I definitely lack the science background to do any sort of meaningful review here! I've a couple of very minor comments.

File: generic_tracers/generic_WOMBATlite.F90 [hunk 4]
```diff
@@ -529,12 +583,12 @@ subroutine generic_WOMBATlite_register_diag(diag_list)
 
     vardesc_temp = vardesc( &
         'htotal', 'H+ concentration', 'h', '1', 's', 'mol/kg', 'f')
```
[the remark this answers is not in the document]
Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:585]: Ah? So what does the L mean and what does the h mean?

File: generic_tracers/generic_WOMBATlite.F90 (hunk 1 above)
Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:1710]: Great catch.

File: generic_tracers/generic_WOMBATlite.F90 (hunk 2 above)
Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:2333]: Also great catch. 

I'm going to update this so I output the feIII, felig and fecol fractions of the total dFe pool.

free iron + ligand-bound iron + colloidal iron == total dFe.

File: generic_tracers/generic_WOMBATlite.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (dougiesquire) [reply generic_tracers/generic_WOMBATlite.F90:585]: Sorry yes, that's a typo from me

File: generic_tracers/generic_WOMBATlite.F90 (hunk 3 above)
Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:1693]: Have removed these unnecessary declarations

**Comment by pearseb** [comment]
Thanks for the review @dougiesquire. Let me know what you think of the updates.

After this, it's the sinking scheme.

File: generic_tracers/generic_WOMBATlite.F90 [hunk 5]
```diff
@@ -2756,6 +2764,18 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
...
+          rmask=grid_tmask, is_in=isc, js_in=jsc, ks_in=1, ie_in=iec, je_in=jec, ke_in=nk)
+
+    if (wombat%id_felig .gt. 0) &
+      used = g_send_data(wombat%id_felig, wombat%felig, model_time, &
+          rmask=grid_tmask, is_in=isc, js_in=jsc, ks_in=1, ie_in=iec, je_in=jec, ke_in=nk)
+
+    if (wombat%id_fecol .gt. 0) &
+      used = g_send_data(wombat%id_fecol, wombat%fecol, model_time, &
+          rmask=grid_tmask, is_in=isc, js_in=jsc, ks_in=1, ie_in=iec, je_in=jec, ke_in=nk)
+
```
Comment (dougiesquire) [inline generic_tracers/generic_WOMBATlite.F90:2778]: You'll need to register these new diganostics with calls to `register_diag_field`

Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90:2778]: Ah. RIght. Done.

File: generic_tracers/generic_WOMBATlite.F90 [hunk 6]
```diff
@@ -757,6 +757,21 @@ subroutine generic_WOMBATlite_register_diag(diag_list)
...
+    vardesc_temp = vardesc( &
+        'felig', 'IIIitation of free Fe onto nanoparticles', 'h', 'L', 's', 'mol/kg', 'f')
+    wombat%id_felig = register_diag_field(package_name, vardesc_temp%name, axes(1:3), &
+        init_time, vardesc_temp%longname, vardesc_temp%units, missing_value=missing_value1)
+
+    vardesc_temp = vardesc( &
+        'fecol', 'IIIitation of free Fe onto nanoparticles', 'h', 'L', 's', 'mol/kg', 'f')
+    wombat%id_fecol = register_diag_field(package_name, vardesc_temp%name, axes(1:3), &
+        init_time, vardesc_temp%longname, vardesc_temp%units, missing_value=missing_value1)
```
Comment (dougiesquire) [inline generic_tracers/generic_WOMBATlite.F90:773]: Should these really all have the same long name?

Comment (pearseb) [reply generic_tracers/generic_WOMBATlite.F90]: Ahhh. In my rush I forgot to do this. Thanks Dougie.

**Review by dougiesquire — APPROVED** [review]
Looks good to me - thanks @pearseb!

**Comment by dougiesquire** [comment]
PS, I'll let you decide whether you want to squash or not. Your commit history looks quite tidy in this one so you could consider selecting "rebase and merge" if you want to keep all the commits

**Comment by pearseb** [comment]
Thanks again Dougie

**Pull request merged**